### DISCUSSION TITLE
Explorer l'historique de la législation

### DISCUSSION THREAD

**Discussion by mquandalle** [betagouv/mon-entreprise#2704] [General]
Actuellement quand la législation évolue, on met à jour les simulateurs, mais on ne peut plus retrouver les anciens taux et relancer une simulation à une date antérieure. C'est une fonctionnalité qu'on a envie d'ajouter depuis un moment sur mon-entreprise (#34), et c'est aussi un besoin pour les utilisateurs de notre future API #1870.

## Étape 1 : « variations »

Pour la mise à jour des règles au 1er janvier 2022, on a conservé les montants de 2021 derrière une condition :

https://github.com/betagouv/mon-entreprise/blob/30bc63bc14b0110faa24114c4d1cfd04d960b3d4/modele-social/r%C3%A8gles/base.yaml#L57-L61

Ça permet déjà de recalculer une fiche de paie avec les valeurs de l'année dernière en ajoutant un paramètre `année: 2021` dans la situation.

Reste à exposer cette fonctionnalité dans l'interface, un simple `<select><option>2021</option><option>2022</option></select>` devrait faire l'affaire. #1138 

## Étape 2 : changement en cours d'année

La législation peut aussi changer en cours d'année, par exemple le SMIC a été augmenté le 1er octobre dernier. On pourrait créer des nouvelles règles à chaque fois pour chaque période mais ça deviendrait rapidement difficile à maintenir. On a besoin d'introduire des notions calendaires du type “à partir de” et “jusqu'au"

```yaml
calendrier:
  - à partir du: 01/01/2022
    valeur: 7.98 €/heure
  - à partir du: 01/10/2021
    valeur: 7.91 €/heure
```

L'intérêt c'est qu'on a plus besoin d'avoir les même “bornes” pour toutes les règles : les dates de changement du SMIC sont indépendantes de celle du plafond de la sécurité sociale ou d'une révision d'une convention collective. Et pour une date donnée on peut toujours déterminer quels dispositifs s'appliquent.

Cela nécessite l'implémentation d'un mécanisme « calendrier » dans publicodes. (discussion historique dans la PR sur les variables temporelles #793)

(edit: en fait non, il est aussi possible de rester avec le mécanisme ` variations` et d'utiliser des comparaison de date `date <= 01/01/2022`)

Côté UI, l'introduction de ce mécanisme peut amener 2 développements :
- afficher un sélecteur de période dans la documentation
- faire de l'inférence de seuils pour proposer un `<select />` avec les dates pertinentes. Par exemple pas besoin d'afficher la date du 1er octobre qui correspond au changement du SMIC sur le simulateur auto-entrepreneur qui n'utilise pas cette valeur dans son calcul. Cette fonctionnalité d'inférences des bornes est implémentée sur mesaidesvelo ([ici](https://github.com/mquandalle/mesaidesvelo/blob/2cb34bea039c669c38532412552f4090ab14dc41/src/lib/components/RevenuSelector.svelte#L6)) et si on l'utilise aussi sur mon-entreprise il deviendra pertinent de l'intégrer dans publicodes https://github.com/betagouv/publicodes/issues/48#issuecomment-973123608

## Étape 3 : découpage de code

Sans découpage de code, on envoie tout l'historique de toutes les règles au client, alors qu'il n'en aura en général pas besoin. Ça rajoute du poids sur le réseau, du temps pour parser tout ça, bref c'est mal.

Là encore on aura besoin de support côté publicodes. Le plus naturel me semble être d'utiliser les fichiers comme unité de découpage. Par exemple pas besoin des règles de l'auto-entrepreneur quand je suis sur le simulateur salarié. À priori ce n'est pas trop compliqué de faire du chargement à chaud de règles avec une API du type `engine.addRules` à condition que l'on s'assure que toutes les dépendances utiles sont bien chargées.

J'ai l'impression que l'on pourrait s'en sortir assez facilement en générant des imports ESM et en délégant à ViteJS la résolution du graphe.

```js
// “compiled” salarie.yaml

import "./base.yaml"

export default {...}
```

S'agissant de l'historique des règles, ça implique aussi de déplacer les anciennes valeurs dans un fichier séparé, au moins si on veut bénéficier des performances du découpage de code. Je pense que ça peut avoir également un intérêt en terme de lisibilité des règles. En effet dans des cas simple comme l'historique du SMIC, la règle reste lisible même avec le mécanisme `calendrier`, mais ce ne sera pas le cas par exemple pour l'historique des barèmes du taux neutre #1959 ([150 lignes par année](https://github.com/betagouv/mon-entreprise/blob/master/modele-social/r%C3%A8gles/imp%C3%B4t.yaml#L105-L258))

Pour pouvoir déplacer l'historique dans un fichier séparé on aura sans doute besoin d'un mécanisme de type “remplace” spécialisé dans la définition d'un historique de formules.

## Étape 4 : échéanciers

Avec le mécanisme calendrier on reste dans l'évaluation d'une situation à un instant donné. L'idée ensuite serait de pouvoir évaluer une période complète en la découpant en mois ou en années. Par exemple on donne une liste de salaires versés où chaque ligne correspond à un mois et on veut prendre en compte les évolutions de la législations intervenus en cours de route pour le calcul des cotisations et impôts.

L'idée a été largement explorée et expérimentée dans la précédente implémentation des « variables temporelles » dans #866, (supprimées dans #1579). Le prérequis est de simplifier l'organisation interne des mécanismes dans publicodes https://github.com/betagouv/publicodes/issues/23

**Comment by mquandalle** [comment]
En fait je ne suis pas sûr de l'intérêt du mécanisme `calendrier` que j'évoquais dans le message ci-dessus, par rapport à l'utilisation de `variations` :

```yaml
date: 17/04/2022

smic:
  variations:
    - si: date >= 01/05/2022
      alors: 1645€
    - si: date >= 01/01/2022
      alors: 1603€
```

[studio](https://publi.codes/studio?code=date%3A+17%2F04%2F2022%0A%0Asmic%3A%0A++variations%3A%0A++++-+si%3A+date+%3E%3D+01%2F05%2F2022%0A++++++alors%3A+1645%E2%82%AC%0A++++-+si%3A+date+%3E%3D+01%2F01%2F2022%0A++++++alors%3A+1603%E2%82%AC%0A)

Inconvénient : ça ressemble un peu plus à une “implémentation” là où la philosophie des mécanismes est plutôt d'exposer des concepts de haut niveau et de garder les modalités de calculs cachées à l'intérieur du mécanisme

Avantage : la dépendance à une variable `date` est explicite, là ou il faudrait l'ajouter implicitement dans un mécanisme `calendrier`. Ça permet notamment de gérer des remplacements sur la date.

**Comment by liliced** [comment]
L'issue #34 a été fait et le découpage de `modele-social` en plusieurs paquets est en cours.
